### PR TITLE
Redirect user to registration flow and introduce them in the activity stream

### PR DESCRIPTION
**Pull request by toolness**
This PR is meant to bludgeon two birds with one stone:

1. Adding the following message to the activity stream when a user has fully completed the registration flow:

  

2. Directing users to complete the registration process if they haven't already (#49).

Technically, this could be split into two separate PRs, but it shouldn't be too hard to do as one.

Tasks left to do:
- [x] Actually create a `SimpleUserEvent.SUBTYPE_USER_JOINED` event when users complete the registration process.
- [x] Add view logic for displaying `SimpleUserEvent.SUBTYPE_USER_JOINED` in the activity stream.
- [ ] Have the `@full_registration_required` decorator redirect the user to step 3 of registration if they've filled out any of the fields in step 2 *or* have already answered any questionnaire questions. (This can potentially be filed as a separate issue and dealt with later.)
- [x] ~~Generate a database migration.~~ [No migration needed now](https://github.com/GovLab/noi2/pull/55#discussion_r40808534)!

### CODE REVIEW

File: app/models.py [hunk 1]
```diff
@@ -437,6 +465,21 @@ class UserEvent(Event):
     def from_user(cls, user, **kwargs):
         return cls(user_id=user.id, deployment=user.deployment, **kwargs)
 
+class SimpleUserEvent(UserEvent):
+    '''
+    Any kind of user event that doesn't have any extra data associated
+    with it.
+    '''
```
Comment (toolness) [inline app/models.py]: This *seems* like a reasonable thing to do based on the mockups, where we have a number of simple events that don't have any extra metadata associated with them, like "A user joined the network" and "A user added expertise".

File: app/models.py [hunk 2]
```diff
@@ -437,6 +473,21 @@ class UserEvent(Event):
...
+
```
Comment (toolness) [inline app/models.py]: Er, note that because we're not specifying a `__tablename__` here, SQLAlchemy will use [single table inheritance](http://docs.sqlalchemy.org/en/rel_0_8/orm/inheritance.html#single-table-inheritance), i.e. the migration will add a column to the table of the superclass (`user_events`) rather than adding a whole new table. Since the only field, `subtype`, is just an integer, this doesn't *seem* like a terrible waste of space, and it also means one less join, but we can change it to use a separate table instead if you want.

File: app/templates/activity.html [hunk 3]
```diff
@@ -46,6 +46,12 @@
     <p>{{ user_link(event.user) }} shared:</p>
 
     {{ event.message }}
+  {% elif event.type == 'simple_user_event' %}
+    {% if event.subtype == event.SUBTYPE_USER_JOINED %}
+      <p>{{ user_link(event.user) }} joined NOI.</p>
```
Comment (toolness) [inline app/templates/activity.html]: Er, this isn't internationalized, but neither are the other messages in the activity stream, yet... Might be easier to just file a separate issue for internationalizing the stream, for now...

**Comment by toolness** [comment]
Note that I'm keeping `login_required` decorators (rather than `full_registration_required`) on `/me` and `/my-expertise`, since those seem like places the user should be able to visit even if they're not fully registered yet, but I can convert those too if needed.

File: app/views.py [hunk 4]
```diff
@@ -22,9 +23,34 @@
...
+    nonetheless meets the criteria for full registration, they will
+    be marked as fully registered and granted access to the view.
+    '''
+
+    @functools.wraps(func)
+    @login_required
+    def decorated_view(*args, **kwargs):
+        if not current_user.has_fully_registered:
+            if current_user.matches_criteria_for_full_registration():
+                current_user.set_fully_registered()
```
Comment (toolness) [inline app/views.py]: Ok, so I'm a bit queasy about this because this decorator is used on plenty of `GET` requests and we're potentially modifying server state, which is generally a no-no.

However, the reason I'm taking this approach is because I realized that there are some weird edge cases with completing registration, e.g.:

1. They could be migrated from a version of a deployment that didn't have registration in-place (though I suppose we could alternatively create a data migration for this).
2. They could answer questions at `/my-expertise` instead of registration step 3 (alternatively, we could require full registration to access that endpoint).
3. There could just be weird edge cases in the future where a user *should* be considered as fully registered but doesn't have the associated event; we don't want to inconvenience users if this happens by needlessly redirecting them to the registration flow.

So, it seemed like the easiest way around this would be to simply add the above check whenever the user tries accessing a view that requires full registration. If this seems gross, though, I can certainly change it to use a different approach.

File: app/models.py (hunk 2 above)
Comment (talos) [inline app/models.py]: No, I think single table inheritance is the right subclassing method to use here.

File: app/templates/activity.html (hunk 3 above)
Comment (talos) [inline app/templates/activity.html]: Yes, please make a separate PR.  Definitely essential as OGP is next big launch.

File: app/models.py [hunk 5]
```diff
@@ -198,6 +198,42 @@ def nearest_neighbors(self, limit=10):
...
+        Returns whether the user has fully completed the registration/signup
+        flow.
+        '''
+
+        join_event = db.session.query(SimpleUserEvent).\
+                     filter(SimpleUserEvent.user_id==self.id).\
+                     filter(SimpleUserEvent.subtype==\
+                            SimpleUserEvent.SUBTYPE_USER_JOINED).\
+                     all()
+        return len(join_event) > 0
```
Comment (talos) [inline app/models.py]: Using `all()` and then `len()` is overkill.  We can just use `.exists()` at the tail end of the query and return that.  See http://docs.sqlalchemy.org/en/latest/orm/query.html?highlight=one#sqlalchemy.orm.query.Query.exists .

File: app/models.py [hunk 6]
```diff
@@ -437,6 +473,21 @@ class UserEvent(Event):
...
+    '''
+
+    # A user completed the registration process and joined the network.
+    SUBTYPE_USER_JOINED = 1
+
+    subtype = Column(types.Integer)
+
+    __mapper_args__ = {
+        'polymorphic_identity': 'simple_user_event'
+    }
```
Comment (talos) [inline app/models.py]: I think it'd be better to do this by having subclasses of `SimpleUserEvent` for each user event type, where the `polymorphic_identity` then gets set to what sort of event it is (for example, `user_joined`).  That way, we can do queries like:

```
UserJoinedEvent.query.filter_by(user_id=<user_id>).exists()
```

To determine whether the user has an event associated without directly querying the subtype.

File: app/views.py (hunk 4 above)
Comment (talos) [inline app/views.py]: Hm.  To reply:


> 1. They could be migrated from a version of a deployment that didn't have registration in-place (though I suppose we could alternatively create a data migration for this).

I think we should handle this explicitly in the migration.

> 2. They could answer questions at /my-expertise instead of registration step 3 (alternatively, we could require full registration to access that endpoint).

I think we should require full registration for them to access that endpoint, especially if that buys us code simplicity.

> 3. There could just be weird edge cases in the future where a user should be considered as fully registered but doesn't have the associated event; we don't want to inconvenience users if this happens by needlessly redirecting them to the registration flow.

While this could be true, it's not yet.  I don't think we need to worry about it.

In short, I think we should equate `login_required` with `full_registration_required`, especially if that buys us code simplicity.

File: app/templates/activity.html (hunk 3 above)
Comment (toolness) [inline app/templates/activity.html]: Sure, filed #58 for this.

File: app/models.py (hunk 6 above)
Comment (toolness) [inline app/models.py]: Er, good idea, that actually even further simplifies things by not requiring a db migration at all!

File: app/models.py (hunk 5 above)
Comment (toolness) [inline app/models.py]: Hmm, I'm trying that but I'm getting `TypeError: Boolean value of this clause is not defined`...

Comment (talos) [inline app/models.py]: I think you'll need to replace `query(SimpleUserEvent)` with `query(SimpleUserEvent.id)`.

File: app/views.py (hunk 4 above)
Comment (toolness) [inline app/views.py]: Sounds good, done!

**Comment by talos** [comment]
:+1:
